docs(): add dark mode theming guide (#542)
ionic-team/ionic-docs: src/pages/theming/dark-mode.md

---
previousText: 'Advanced'
previousUrl: '/docs/theming/advanced'

> contributors:

Reviewer: Shouldn't there be a next page here still? Notice advanced goes on to publishing.

I think it should actually switch order though so dark mode is before advanced.

Author: My though is that since dark mode is kind of an advanced feature, it should come after. We can add next buttons and send it to publishing from dark mode?

  - mhartington
---

> # Dark Mode

Reviewer 2: Add a reference to the `css/themes` directory where they can find a few examples

> Dark Mode is a specific theme that developers can add to their app in two ways. There's the new `prefers-color-scheme` media query or the fallback method of targeting a class. To create a dark mode for an app, all that's needed are a few changes to different CSS variables.

Reviewer: ```suggestion
Dark mode is a specific theme that developers can add to their app in two ways. Developers can use the new `prefers-color-scheme` media query or use the fallback method of targeting a class. To add a dark mode version of an app, all that's needed are a few custom CSS variables.
```

## Using Media Queries

The first method available is the CSS media query for the user's preferred color scheme. This media query will hook into the system setting of the user's device and apply the changes if a dark mode is enabled.

> The fist values that will need to change are the text color and background color. As the name "dark mode" implies, the background of an app should be dark, with lighter text.

Reviewer: ```suggestion
The first values that need to change are the text color and background color. As the name "dark mode" implies, the background color of an app should be dark with a lighter text color.
```

```css
@media (prefers-color-scheme: dark) {
  :root {

>     --ion-background-color: var(--ion-color-dark);

Author: Mmm, figured using the variables here would be better as a way to tie things back to the main css theme.

>     --ion-background-color: var(--ion-color-dark);

Reviewer: True, maybe link to the description of the Ionic colors then? https://ionicframework.com/docs/theming/advanced#colors

In addition, the step variables can also be modified to further enhance the dark mode experience.

```css
@media (prefers-color-scheme: dark) {
  :root {
    --ion-background-color: var(--ion-color-dark);
    --ion-background-color-rgb: var(--ion-color-dark-rgb);

    --ion-text-color: var(--ion-color-light);
    --ion-text-color-rgb: var(--ion-color-light-rgb);

    --ion-color-step-50: #2d2e32;
    --ion-color-step-100: #37393d;
    --ion-color-step-150: #424347;
    --ion-color-step-200: #4c4e52;
    --ion-color-step-250: #57585c;
    --ion-color-step-300: #616366;
    --ion-color-step-350: #6c6d71;
    --ion-color-step-400: #76787b;
    --ion-color-step-450: #818286;
    --ion-color-step-500: #8b8d90;
    --ion-color-step-550: #96979a;
    --ion-color-step-600: #a0a1a5;
    --ion-color-step-650: #abacaf;
    --ion-color-step-700: #b5b6ba;
    --ion-color-step-750: #c0c1c4;
    --ion-color-step-800: #cacbce;
    --ion-color-step-850: #d5d6d9;
    --ion-color-step-900: #dfe0e3;
    --ion-color-step-950: #e9ebee;
  }
}
```

> In this case, the color steps are generated based on the background color and text color. The [stepped color generator](/docs/theming/advanced#generate-stepped-color-variables) can take the hex value of the background color and automatically generate the steps needed for an app.

Reviewer: We should move this above the code so it isn't missed. 🙂

> In this case, the color steps are generated based on the background color and text color. The [stepped color generator](/docs/theming/advanced#generate-stepped-color-variables) can take the hex value of the background color and automatically generate the steps needed for an app.

Reviewer: ```suggestion
In this case, the color steps are generated based on the background color and text color. The [stepped color generator](/docs/theming/advanced#generate-stepped-color-variables) can take the hex value of the background color and text color and automatically generate the steps needed for an app.
```

## Class Fallback

> Currently, the `prefers-color-scheme` media query is only supported in desktop Safari and mobile Firebox, so users will not able to benefit from having the dark mode applied automatically. But the dark mode can still be applied by using a CSS selector and applying a class to our document body.

Reviewer: ```suggestion
Currently, the `prefers-color-scheme` media query is only supported in desktop Safari and mobile Firebox, so users will not able to benefit from having the dark mode applied automatically in other browsers. However, the dark mode can still be applied by styling a CSS selector and applying the class to our document body.
```

> Currently, the `prefers-color-scheme` media query is only supported in desktop Safari and mobile Firebox, so users will not able to benefit from having the dark mode applied automatically in other browsers. However, the dark mode can still be applied by styling a CSS selector and applying the class to our document body

Reviewer: ```suggestion
Currently, the `prefers-color-scheme` media query has [limited browser support](https://caniuse.com/#feat=prefers-color-scheme), so users will not be able to benefit from having the dark mode applied using this media query in certain browsers. However, the dark mode can still be applied by styling a CSS selector and applying the class to our document body
```

Maybe here we should just link to the browser support? That way we don't have to update the docs as support grows.
